### PR TITLE
feat(prefer-primordials): Add check for getter properties

### LINKED ISSUE #1112: prefer-primordials: Add check for getter properties
```js
// call ArrayBuffer.prototype.byteLength getter function
new ArrayBuffer(10).byteLength;
```

should be changed to the following:

```js
ArrayBufferPrototypeGetByteLength(new ArrayBuffer(10));
```

---

Since type information is not available, it is the only way to check if the property names are defined as getter properties in ECMAScript:

`Symbol`

* `description`

`RegExp`

* `dotAll`
* `flags`
* `global`
* `hasIndices`
* `ignoreCase`
* `multiline`
* `source`
* `sticky`
* `unicode`

`TypedArray`

* `buffer`
* `byteLength`
* `byteOffset`
* `length` 👉 better to ignore it because it cause false positives for the `Array`s

`Map`, `Set`

* `size`

`ArrayBuffer`, `SharedArrayBuffer`

* `byteLength`

`DataView`

* `buffer`
* `byteLength`
* `byteOffset`

### PR DESCRIPTION
**Pull request by petamoriken**
Closes #1112 

Since I reordered the functions in the second commit, it is easy to check the diffs on a per-commit base.

### CODE REVIEW

**Comment by petamoriken** [comment]
@magurotuna please review

File: src/rules/prefer_primordials.rs [hunk 1]
```diff
@@ -261,15 +229,29 @@ impl Handler for PreferPrimordialsHandler {
...
+      if let Expr::Ident(ident) = &member_expr.obj;
+      if GLOBAL_TARGETS.contains(&ident.sym().as_ref());
+      then {
+        ctx.add_diagnostic_with_hint(
+          member_expr.range(),
+          CODE,
+          PreferPrimordialsMessage::GlobalIntrinsic,
+          PreferPrimordialsHint::GlobalIntrinsic,
+        );
+        return;
```
Comment (petamoriken) [inline src/rules/prefer_primordials.rs:245]: Should I remove this `return`?

**Review by magurotuna — COMMENTED** [review]
Maybe I'm missing something, but code like `ourOwnInstance.buffer` will be warned with this new feature, right? Is it intentional?
Although there's no type information available, just checking potentially dangerous getters by matching its name looks a bit too restrictive.

**Comment by petamoriken** [comment]
> Maybe I'm missing something, but code like ourOwnInstance.buffer will be warned with this new feature, right? Is it intentional?

Yes.

> Although there's no type information available, just checking potentially dangerous getters by matching its name looks a bit too restrictive.

This way, it is hard to avoid false positives, but since most of the time `.buffer` is accessed by `TypedArray` or `DataView`, it is unlikely to cause any problems. If false positives are detected, we can still use `// deno-lint-ignore prefer-primordials` to work around them.

**Comment by magurotuna** [comment]
Thanks for clarification. I get the point and am inclined towards it, but still not 100% sure if it's worth the effort and cost. (Also a bit worried if this might have adverse impacts on performance)

@bartlomieju Any thoughts? Do you think that protecting getters by matching its name us is worthwhile?

**Comment by bartlomieju** [comment]
> Thanks for clarification. I get the point and am inclined towards it, but still not 100% sure if it's worth the effort and cost. (Also a bit worried if this might have adverse impacts on performance)
> 
> @bartlomieju Any thoughts? Do you think that protecting getters by matching its name us is worthwhile?

It's okay to me as long as the changes prompted by this lint don't cause performance regressions.

**Comment by magurotuna** [comment]
OK thanks, so let's land it and see how the performance changes.

File: src/rules/prefer_primordials.rs [hunk 2]
```diff
@@ -637,6 +670,17 @@ const noop = Function.prototype;
...
+const { ArrayBuffer } = primordials;
+new ArrayBuffer(10).byteLength;
+      "#: [
+        {
+          line: 3,
+          col: 0,
+          message: PreferPrimordialsMessage::GlobalIntrinsic,
+          hint: PreferPrimordialsHint::GlobalIntrinsic,
+        },
+      ],
```
Comment (magurotuna) [inline src/rules/prefer_primordials.rs:683]: Could you also add other test cases to check all the getters included in `GETTER_TARGETS` other than `byteLength`? I think it should be helpful to make sure everything is working

Comment (petamoriken) [reply src/rules/prefer_primordials.rs:762]: Done!

**Comment by petamoriken** [comment]
I Updated to avoid false positives for setter properties.

File: src/rules/prefer_primordials.rs [hunk 3]
```diff
@@ -438,6 +474,8 @@ const parseInt = () => {};
 parseInt();
       "#,
       r#"const foo = { Error: 1 };"#,
+      r#"foo.size = 1"#,
+      r#"foo.size()"#,
```
Comment (magurotuna) [inline src/rules/prefer_primordials.rs:478]: 👍

**Review by magurotuna — APPROVED** [review]
Great, LGTM

**Pull request merged**